### DISCUSSION TITLE
MinIO process gets killed by FreeBSD under load (too many threads)

### DISCUSSION THREAD

**Discussion by mrdsam** [minio/minio#12675] [Q&A] [answered]
MinIO process gets killed by OS after spawning more than <kern.threads.max_threads_per_proc> (1500 or 2000) threads under high load (writing small objects).

## Expected Behavior
Process should not spawn this many threads, instead deny service to client temporarily.

## Current Behavior
Minio process spawns processes until it gets killed by FreeBSD

## Possible Solution
Implement a max_threads option or detect limits automatically

## Steps to Reproduce (for bugs)
Build a 4-Node-Cluster with FreeBSD 12.2 with very slow disks and write a lot of small objects simultaneously.

## Context
Cluster not stable enough for production

## Your Environment

* minio version RELEASE.2021-03-17T02-33-02Z
* FreeBSD 12.2-RELEASE-p7 GENERIC  amd64
* 4 Nodes, each with 64 GB RAM, 4x 16TB HDD, 1 Gbit NIC, one ZFS-Pool per HDD

**Comment by harshavardhana** [comment, marked as answer]
@mrdsam please upgrade to latest release and let us know how it goes. 

> MinIO process gets killed by OS after spawning more than <kern.threads.max_threads_per_proc> (1500 or 2000) threads under high load (writing small objects).

If you have higher incoming requests you need to increase this value yourself, an arbitrary limit like this and pushing the server beyond this limit will definitely crash the server. 

NOTE: we do not manage the threads, this is up to Go runtime - so whatever it thinks that it needs a thread it shall create on the system for filesystem I/O.

**Comment by mrdsam** [comment]
Thanks for quick response. At the moment, I can not upgrade. RELEASE.2021-03-17T02-33-02Z is the latest **working** port as minio-2021.04.06.23.11.00 is not usable due to bug #12349.

Unfortunately, I haven't learned how to package ports so I depend on the port maintainer.

Another question in this context: Is MinIO using the /proc filesystem? Because it runs without, but sometimes, in rare cases, there are "can not find /proc/self" messages on the console. I mounted /proc  eventually, but don't know if this is related.

**Comment by harshavardhana** [comment]
> Another question in this context: Is MinIO using the /proc filesystem? Because it runs without, but sometimes, in rare cases, there are "can not find /proc/self" messages on the console. I mounted /proc eventually, but don't know if this is related.

It is only for situations if you are monitoring or running some prometheus metrics we require access to `/proc` also that information are only supported primarily on Linux.
